Add index length option for MySQL

abeverley opened the pull request:
This adds an option to be able to specify index lengths for MySQL, such as:

`$index->fields( 'id', { name => 'firstname', size => 15 } );`

The majority of this PR is updating the other providers to ignore this, and a bunch of tests (including a new basic SQLServer test).

castaway commented:
I like the idea, I'd be tempted to try and hide the fact that the fields may be either strings or hashrefs more... that is instead of having ref $_ ? everywhere, make a method field_names on the Index object which just returns the strings, and maybe another one fields_with_lengths, which returns names and/or name(size) strings as appropriate.

That way if we at some point add more attrs or just make them all refs, we only change one place.

ribasushi commented:
@castaway thanks for looking - I agree with the *please make a helper function* part.

@abeverley Recall how I was unsure about the name - after reading [various manuals](http://dev.mysql.com/doc/refman/5.7/en/create-index.html) I am now pretty sure `size` would be a mistake, as it is pretty vague. Would you be opposed to renaming the option to `prefix_length`: this way it is both much clearer **and** much more to type, to make a user think (few engines, even within MySQL itself support this, namely only `innodb`). Additionally there are things like `key_block_size` which one may want to specify as well some day, and having already a `size` will be mighty confusing.

I am volunteering to do the renaming work (you've waited long enough on this PR) - I am just making sure you are not opposed to the changes (as you will need to reflect them on your end :/)

ribasushi commented:
@ilmari, @mcsnolte: just a heads up that this one needs to go in for next rel, thus delaying things by several days. I was under the impression someone merged this already, my bad entirely :/

mcsnolte commented:
@ribasushi got it, thanks for the update!

abeverley commented:
On Wed, 2015-11-04 at 06:14 -0800, Peter Rabbitson wrote:
> Would you be opposed to renaming the option to `prefix_length`

Not at all, absolutely fine.

> I am volunteering to do the renaming work (you've waited long enough 
> on this PR)

Fantastic, thanks.

Andy

abeverley commented:
On Wed, 2015-11-04 at 03:12 -0800, castaway wrote:
> I like the idea, I'd be tempted to try and hide the fact that the 
> fields may be either strings or hashrefs more... that is instead of 
> having ref $_ ? everywhere, make a method field_names on the Index 
> object which just returns the strings, and maybe another one 
> fields_with_lengths, which returns names and/or name(size) strings as 
> appropriate.

Thanks @castaway, makes sense. I don't want to hold a release up though,
and I'm out all day today, so would you mind doing that as part of my
Shadowcat ticket please? (assuming you have time today).

Thanks,

Andy

ribasushi commented:
@castaway Leaving the part @abeverley requested to you unless I hear otherwise...

castaway commented:
@abeverley @ribasushi  yes, can do (tho @ribasushi may have to remind me of the best way to add to/update the PR)

ribasushi commented:
@castaway Don't worry about *proper GitHub etiquette* (as if such a thing ever existed). Get his changes in some form, mangle them, then push a branch of your own naming to the main repo. I will pick it up from there.

castaway commented:
Fixes pushed to branch on my github clone with same name.. (and @ribasushi poked with it)

abeverley commented:
I've just realised that I don't think this ever got merged. I realised by building a new dev machine and finding the use of this feature not working, so would be good to prevent that happening again in the future ;-)

Would it be possible to merge it please? I've merged in @castaway's changes as discussed above and fixed some merge conflicts. Thanks!

abeverley commented:
@rabbiveesh - I understand that you have recently taken over this module. Would there be any chance of merging this long-standing PR please? I have updated it to allow merging with the latest version of master. I have to keep patching my installations, so would appreciate it being in the CPAN release! Many thanks.

abeverley commented:
@rabbiveesh @nrdvana ping - any chance of merging this please? I've just rebased to master again, which takes a while each time I have to do it, so would be grateful not to have to do it again :-) Thanks.

nrdvana commented:
@abeverley (not a maintainer, but...) does this patch effectively add the ability to put arbitrary attributes onto a field in an index?  If so, that would go a long ways to solving my problem in #154, so I'm fairly interested in this now.

Could you quickly summarize the changes here (like bullet list for changelog style) and whether it breaks any compatibility with previous versions?

abeverley commented:
@nrdvana - yes, no problem at all. I think this patch would give you the foundation to implement #154. It enables fields to be specified as hashrefs (with properties) rather than just a simple scalar name. This particular patch then adds only one property, which is ```prefix_length```, but all the hard work is done so you could probably add other properties (I have just pushed another commit now that should enable that to happen, so that MySQL will ignore the option if it's not ```prefix_length```).

In changelog style:

- Enable index lengths to be specified for each indexed field in MySQL

As an example:

```$sqlt_table->add_index(name => 'user_idx_value', fields => [ { name => 'value', prefix_length => 64 } ]);```

So you could add another property, e.g.

```$sqlt_table->add_index(name => 'user_idx_t', fields => [ { name => 't', opclass => 'gin_trgm_ops' } ]);```

Take a look at this for example code which you could add to the PostgreSQL producer: https://github.com/abeverley/sql-translator/blob/d064562a9ff1f0f227a6bcd0ec8c6242454d6d61/lib/SQL/Translator/Producer/MySQL.pm#L688

abeverley commented:
@nrdvana and no it doesn't break anything, this is backwards-compatible. If the index is a scalar then it uses it as normal, if it's a hashref then it pulls out the options. If @rabbiveesh can merge it, then you can build a new option ```opclass``` or similar quite easily.

rabbiveesh commented:
Yeah, i don't really like this approach so much; index prefix length is a database dependent feature - the handling for that should be in the databases that support it, not in the abstract class.

i also don't like the fact that this change bleeds everywhere all over the code; i would go further than @nrdvana and say that `fields` should continue returning just the name of the field, and there should be a separate `field_info` method or something for that auxiallary info.

In fact, `field_info` or whatever should always return hashes, so we never have to deal with mixed lists.

I'm gonna give this a crack

rabbiveesh commented:
now that i'm hacking on this, i like the idea that `fields` will start returning objects with a stringification overload. Otherwise, it would involve ugly things on the inside of the Index class, so it seems like that would be the nicest option.

Other thoughts: index options currently use a `extras` hash to store producer specific data. We could do the same for field options (I would have the object representing fields parse anything other than `name` into the `extras` hash.

abeverley commented:
Thanks @nrdvana and @rabbiveesh for your thoughts - they all make sense. The code that is there is what came out of similar discussions several years ago (I'm not sure why it was never merged originally, I think it just got lost), but I am more than happy to revisit the concept with your fresh pairs of eyes.

Would you like me to work on any of the above?

rabbiveesh commented:
I'm writing up something now, I'm gonna keep your tests and do it with new IndexFields objects

nrdvana commented:
@rabbiveesh As you design that, an interesting unit test for postgres to really test the boundaries would be

```
create table test (name varchar(50) not null);
create index ix_test2 on test (substr(name, 2, 3) desc, substr(name, 5, 3) asc);
```

:-)

rabbiveesh commented:
> @rabbiveesh As you design that, an interesting unit test for postgres to really test the boundaries would be
> 
> ```
> create table test (name varchar(50) not null);
> create index ix_test2 on test (substr(name, 2, 3) desc, substr(name, 5, 3) asc);
> ```
> 
> :-)

Can you put this on the other ticket?
I think i'm gonna do the PG side of this in a different PR, mainly b/c I haven't settled on how mch work to make the parser do

nrdvana commented:
Well, to spell out the thought I had, the column of an index isn't always a column; it could be a function.  And there could be two functions of the same column, so the unique check probably needs to be removed.  And (I still didn't check this) if existing parsers are returning sql expressions as the 'field' currently, then the thing that the object returns when stringified should probably be an attribute named something like 'original_sql_expr' instead of 'name'.  Quoting the name in a producer should probably only happen if it is known to be a simple name.  And all these things probably need to be thought about now in order to make the postgres parser/producer reasonable to implement.

rabbiveesh commented:
I think that the design we settled on is sufficiently encapsulated that we can make changes later.
My first thought is that to handle where the field is really an expression, we should have an `expression` flag in the `extra` hash, which the postgres producer will then use to skip quoting even when it's on. Hopefully it won't be too hard to recognize an expression, I haven't worked on the parsers in a while.

re `uniq` - in my implemenation, i actually removed that, b/c the `fields` of an `Index` need to be coerced into the new `IndexField` objects, so that's handled there. Though i see the logic in not allowing duplicates (allows calling code to be lazy in programatically adding column names or whatever).

rabbiveesh commented:
The truth is, using expressions as index fields would actually need it's own support completely, b/c it would break the very naive `is_valid` check on `::Index`.

We should write up all these notes in a new ticket (*hint hint*)

rabbiveesh commented:
Fine, i opened the other ticket

rabbiveesh commented:
@abeverley I have opened the PR abeverley/sql-translator#1 with my changes. Please review it w/ your thoughts and/or add to its completeness.

If you could add the `extra` key to the `YAML` and `JSON` parser etc in the event that it has data, that would be awesome

abeverley commented:
> If you could add the `extra` key to the `YAML` and `JSON` parser etc in the event that it has data, that would be awesome

Ah, I just saw this comment after posting my other one. I will merge your changes and then add the key as suggested.

abeverley commented:
Okay, I've just done that (https://github.com/dbsrgits/sql-translator/pull/68/commits/bfc7565e884e646ee93378ec2428e77bea2c0164). I've only done it in the YAML producer, as the JSON one seems to be structured differently and doesn't return a list of fields. I noticed that I didn't touch the JSON producer originally, possibly because of that limitation. The YAML producer was changed in this commit (4a06a63f38d6ff6436e9b18bbec405faa7d092a5) - I am not sure if that should also be applied to the JSON producer?

rabbiveesh commented:
Super sorry that I've dropped the ball on this one. 

Where are we holding?

abeverley commented:
> Super sorry that I've dropped the ball on this one.
> 
> Where are we holding?

No problem at all. Personally I think it is all good to go. I've used your revised implementation (much better), made some minor updates and fixed the tests. The only thing I haven't done is update the JSON producer, but that's because it seems to not be applicable so there is nothing to update.

Let me know if you disagree :)

rabbiveesh commented on t/13schema.t:
```diff
@@ -248,10 +248,45 @@ require_ok( 'SQL::Translator::Schema' );
+
     my $indices = $person_table->get_indices;
-    is( scalar @$indices, 2, 'Two indices' );
+    is( scalar @$indices, 3, 'Two indices' );
```
oops, gotta fix that test name

rabbiveesh commented on lib/SQL/Translator/Producer/YAML.pm:
```diff
@@ -173,7 +173,8 @@ sub view_index {
         'name'      => scalar $index->name,
         'type'      => scalar $index->type,
-        'fields'    => [ map { ref($_) ? $_->name : $_ } $index->fields ],
+        # If the index has extra properties, make sure these are written too
+        'fields'    => [ map { ref($_) && $_->extra && %{$_->extra} ? { name => $_->name, %{$_->extra} } : "$_" } $index->fields ],
```
we should make sure we do the same for JSON

rabbiveesh reviewed (changes requested):
i'm gonna do the last touches here

rabbiveesh commented:
@abeverley made a PR against your PR again, merge it up + we're good to go over here

abeverley commented:
Brilliant, thanks @rabbiveesh - really pleased we've got this merged now. It was definitely worth spending a bit of time on it. As a final test I've just tried it on my local code base and I can confirm it works as expected.
